File: src/main/java/me/miki/shindo/utils/Multithreading.java
package me.miki.shindo.utils;

import java.util.concurrent.*;

public class Multithreading {

    private static final ScheduledExecutorService RUNNABLE_POOL = Executors.newScheduledThreadPool(8, Thread::new);

    public static ExecutorService POOL = Executors.newCachedThreadPool(Thread::new);

    public static void schedule(Runnable r, long initialDelay, long delay, TimeUnit unit) {
        RUNNABLE_POOL.scheduleAtFixedRate(r, initialDelay, delay, unit);
    }

    public static void schedule(Runnable r, long delay, TimeUnit unit) {
        RUNNABLE_POOL.schedule(r, delay, unit);
    }

    public static int getTotal() {
        return ((ThreadPoolExecutor) Multithreading.POOL).getActiveCount();
    }

    public static void runAsync(Runnable runnable) {
        POOL.execute(runnable);
    }
}

File: src/main/kotlin/me/miki/shindo/utils/Multithreading.kt
package me.miki.shindo.utils

import me.miki.shindo.utils.concurrent.TaskExecutor
import me.miki.shindo.utils.concurrent.TaskPriority
import me.miki.shindo.utils.concurrent.ThreadPoolType
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

@Deprecated(
    "Use TaskExecutor for better control and features",
    ReplaceWith("TaskExecutor.runAsync(ThreadPoolType.GENERAL) { }")
)
object Multithreading {

    @JvmStatic
    var POOL = TaskExecutor

    @JvmStatic
    fun schedule(runnable: Runnable, initialDelay: Long, delay: Long, unit: TimeUnit) {
        TaskExecutor.scheduleAtFixedRate(ThreadPoolType.SCHEDULED, initialDelay, delay, unit, runnable)
    }

    @JvmStatic
    fun schedule(runnable: Runnable, delay: Long, unit: TimeUnit) {
        TaskExecutor.schedule(ThreadPoolType.SCHEDULED, delay, unit, runnable)
    }

    @JvmStatic
    fun getTotal(): Int {
        val stats = me.miki.shindo.utils.concurrent.ThreadPoolManager.getAllStats()
        var total = 0
        for (stat in stats.values) {
            total += stat.activeCount
        }
        return total
    }

    @JvmStatic
    fun runAsync(runnable: Runnable) {
        TaskExecutor.runAsync(ThreadPoolType.GENERAL, TaskPriority.NORMAL, runnable)
    }
    fun runAsync(task: () -> Unit): CompletableFuture<Void> {
        return TaskExecutor.runAsync(ThreadPoolType.GENERAL, TaskPriority.NORMAL, Runnable { task() })
    }
}
